src/runtime.rs

```diff
@@ -650,3 +594,120 @@ impl<DB: Database> DependencyGraph<DB> {
         }
     }
 }
+
+/// The `Frozen` struct indicates a database handle which is locked at
+/// a particular revision: any attempt to set the value of an input
+/// (e.g., from another database handle) will block until the `Frozen`
+/// database is dropped.
+///
+/// # Deadlock warning
+///
+/// Since attempts to set inputs are blocked until the `Frozen<DB>` is
+/// dropped, this implies that any attempt to set an input *using* the
+/// `Frozen<DB>` will deadlock instantly (because it will block the
+/// thread that owns the `Frozen<DB>` and thus not permit the
+/// `Frozen<DB>` to be dropped). In the future, we plan to refactor
+/// the API to make such "instant deadlocks" impossible.
+pub struct Frozen<DB>
+where
+    DB: ParallelDatabase,
+{
+    shared_state: Arc<SharedState<DB>>,
+    db: DB,
+}
+
+impl<DB> Frozen<DB>
+where
+    DB: ParallelDatabase,
+{
+    /// Creates and returns a frozen handle to `source_db`.
+    pub(crate) fn new(source_db: &DB) -> Self {
+        let source_runtime = source_db.salsa_runtime();
+
+        // Subtle point: if the source database is already executing a
+        // query, then we want to use a "recursive read" lock.  Using
+        // an ordinary read lock [may deadlock], since it could wind
+        // up blocking indefinitely if there is a pending write (thus
+        // preventing our caller from releasing their read lock, which
+        // in turn ensures that the pending write will never
+        // complete).
+        //
+        // We could just use recursive reads *always*, but that could
+        // lead to starvation in the case where you have various
+        // threads invoking get and set willy nilly. Not sure how
+        // important it is to ensure that case works -- it's not a
+        // recommended pattern -- but it seems (for now at least) easy
+        // enough to do so.
+        //
+        // [may deadlock]: https://docs.rs/lock_api/0.1.3/lock_api/struct.RwLock.html#method.read
+        let use_recursive_lock = source_runtime.query_in_progress();
```

Comment:
This indeed looks subtle and too elaborate for a thing which is basically a counter for "how many forks are there". A particular funny point is that, if `use_recursive_lock` is `true`, then we **know** that we already hold the lock, so we don't really need to lock it, we need to upgrade the locked counter. That is, we know that we can't block.

I feel like there's perhaps a higher-level synchronization primitive we can use here, but I can't think of one. Perhaps a frozen runtime can hold an `Arc<Thing>`, such that forking a frozen runtime just bumps ref count on the `thing`, and, when the last runtime dies, the thing's drop unlocks some lock?

Reply:
>  A particular funny point is that, if `use_recursive_lock` is true, then we know that we already hold the lock, so we don't really need to lock it, we need to upgrade the locked counter

The intuition *here* I think is that -- if this flag is true -- we are "splitting" our existing read permission into two (hence we acquire `read_recursive`, to do that). The main use case I see for this is that we will want to permit one query to fork off new (temporary) handles in order to do things like "do-all-queries-in-parallel", so it may want to invoke `fork()` (it is interesting that it can also invoke `fork_mut`, although it doesn't violate any of the nice properties -- kind of a foot gun though).

> I feel like there's perhaps a higher-level synchronization primitive we can use here, but I can't think of one.

I know what you mean, but I too couldn't think of one. I did consider making our own primitive here -- basically wrapping a `Mutex<usize>` and a condition variable -- but it seems like a reasonable amount of logic to reproduce for little reason.

An alternative is just to **always** use `read_recursive`. This is not terrible. The main problem is that if you have two threads just doing random gets/sets, the `set` operations may starve. This wouldn't really arise in the "expected usage pattern", though, because there we have one master thread doing the sets *and* giving rise to the read-only queries, and those read-only queries always execute with a frozen DB anyway.